### ISSUE TITLE
Self-sizing cells crash bug

### ISSUE THREAD

**Issue by PhilCai1993** [Instagram/IGListKit#260]
### General information

- `IGListKit` version: latest on master
- iOS version(s): iOS 9.3.1 (jailbroken)
- Xcode version: 8.1 from App Store
- Devices/Simulators affected: iPhone 6
- Reproducible in the demo project? Yes

![image](https://cloud.githubusercontent.com/assets/6828315/20617884/e5b6b046-b326-11e6-89ae-efaed2751ffd.png)

![image](https://cloud.githubusercontent.com/assets/6828315/20617887/f2269e72-b326-11e6-9adb-4000e5c4032d.png)

**Comment by PhilCai1993** [comment]
After checking [StackOverFlow](http://stackoverflow.com/questions/25895311/uicollectionview-self-sizing-cells-with-auto-layout/25896386)
This can be solved 
```
final class ManuallySelfSizingCell/FullWidthSelfSizingCell/NibSelfSizingCell: UICollectionViewCell {
      var hasCalculated = false
      ......
      override func preferredLayoutAttributesFitting(_ layoutAttributes: UICollectionViewLayoutAttributes) -> UICollectionViewLayoutAttributes {
        if hasCalculated {
          return layoutAttributes
        }
        hasCalculated = true
        setNeedsLayout()
        layoutIfNeeded()
        let size = contentView.systemLayoutSizeFitting(layoutAttributes.size)
        var newFrame = layoutAttributes.frame
        // note: don't change the width
        newFrame.size.height = ceil(size.height)
        layoutAttributes.frame = newFrame
        return layoutAttributes
    }
}
```
```ManuallySelfSizingCell``` and ```FullWidthSelfSizingCell``` can be solved on iOS 9.
But ```NibSelfSizingCell``` behaves a little strange.
@rnystrom

**Comment by jessesquires** [comment]
Thanks for the update @PhilCai1993 !

**Comment by jeffbailey** [comment]
I'm currently using self sizing cells with IGListKit in a client app and had this crash.  I tracked it down to unneeded calls to:

`setNeedsLayout()`
` layoutIfNeeded()`

in preferredLayoutAttributesFitting.  I wasn't able to reproduce the issue in the demo app, but you might want to try removing those methods.

I think it makes sense that these calls aren't needed by systemLayoutSizeFitting to calculate the size.

**Comment by rnystrom** [comment]
@jeffbailey good to know. I should try this out on iOS 9/10, maybe even put a version check in there if the behavior changes. Did you hit this crash on 10?

FYI I pasted the solution from http://stackoverflow.com/a/25896386/940936 and removed the caching bit that @PhilCai1993 called out as fixing iOS 9 b/c I ran into some weird behavior on 10.

**Comment by jeffbailey** [comment]
Yes, the crash was on 10, but went away as soon as I removed those lines.  I believe it had to do with those methods being called on cells that were about to be reused, but that's just a guess.

**Comment by rnystrom** [comment]
@jeffbailey if you run into it again or are able to make the example project do this I'd love to take a deeper look. I'll play around w/ different setups to see what works best here in the meantime.

**Comment by sun-fox-cj** [comment]
How to solve this crash on 10?

**Comment by rnystrom** [comment]
@sun-fox-cj have you given @jeffbailey's solution a try?

**Comment by sun-fox-cj** [comment]
I have solved this problem without version check. Before I fix this crash, I did it as follows:

1.  First, add a width constraint equal to screen's width on contentView, 
2. and then fill those UIControls which in contentView with the model's data.
3. Next step is that invoke `layoutIfNeeded ` and `systemLayoutSizeFitting` method to calculate the height of content view, then cache it. 

It works well before iOS10, `but` crashes in iOS10. 

So I change it like this: I add inner UIControls constraint with contentView to certain the width of contentView instead of add a width constraint directly. It works well, again.

**Comment by WeiweiYuan810** [comment]
any update?

**Comment by shuhrat10** [comment]
@rnystrom I'm getting a jerky animation when using the searchBar and self-sizing cells. Have you seen this before?

**Comment by shuhrat10** [comment]
I'm still getting crash on iOS 9.3 in Example iOS App (Self-sizing cells).
